### DISCUSSION TITLE
How to update large amounts of text without blocking UI

### DISCUSSION THREAD

**Discussion by yggdr** [Textualize/textual#5057] [Q&A]
While playing around with what textual can do I have built a very simple repository updater, that keeps my source code clones up to date and shows me what's new in a few projects (like e.g. textual).
Works fine so far, it starts up, runs various `git pull` in background-processes and shows the output in a `Static`. If I see something that interests me, I can then have a simple `git diff` run in the background and call `Static.update` to now render the diff. The problem I have is that `Static.update` lags immensely once we get to text sizes in the megabytes, which for a git diff of a medium sized project can happen rather quickly. That lag completely blocks the UI for several seconds, as `rich.segment.split_and_crop_lines` seems to take very long, at least from a cursory investigation via flamegraph.

So what I'm asking is: what is the best approach to such a large-size update in the textual ecosystem if `Static.update` is not it?

**Comment by TomJGooding** [comment]
Have you looked at using the [Worker API](https://textual.textualize.io/guide/workers)?

**Reply by yggdr** [reply]
Using a worker (with `thread=True` or not) doesn't really change much. The UI interruption seems a little less extreme because a few changes are visible earlier. But the UI still does not react to input for essentially the same amount of time.

**Comment by yggdr** [comment]
I have replaced `Static` with a `Log`, which seems to be much faster, although still blocking. This might have to do with the [linewise API](https://textual.textualize.io/guide/widgets/#line-api) that I am only just reading up about.

**Comment by willmcgugan** [comment]
Updating Static with megabytes of data will certainly prompt a lot of work. We're talking hundreds of pages of text being reflowed each update.

Using a Log will updating a line at a time. But you can still freeze the UI if you write many lines in an event handler.

The solution probably involves workers, but I couldn't say for certain without seeing your code. Since you are using a lot of concurrency, you may be blocking without realizing it. And that may not be obvious from profiling.
